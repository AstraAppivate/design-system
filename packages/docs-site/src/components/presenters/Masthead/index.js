import React, { useState } from 'react'
import {
  Button,
  Nav,
  PhaseBanner,
  Icons,
} from '@royalnavy/react-component-library'
import { Form, Field, Formik } from 'formik'

import TextInput from '../TextInput'
import SiteLogo from './images/site-logo.svg'

import './masthead.scss'

const { Search, TriangleDown, TriangleUp } = Icons

const MastHead = () => {
  const [open, setOpen] = useState(false)

  const initialValues = { search: '' }

  const onSubmit = term => {
    console.log('Do a thing:', term)
  }

  const toggle = () => {
    setOpen(!open)
  }

  const navItems = [
    {
      href: '1',
      label: 'Get started',
    },
    {
      href: '1',
      label: 'Styles',
    },
    {
      active: true,
      href: '1',
      label: 'Components',
    },
    {
      href: '1',
      label: 'Patterns',
    },
    {
      href: '1',
      label: 'Community',
    },
    {
      href: '1',
      label: 'About',
    },
  ]

  return (
    <div className="masthead">
      <div className="masthead__container">
        <SiteLogo className="masthead__logo" />
        <Formik initialValues={initialValues} onSubmit={onSubmit}>
          <Form className="masthead__search">
            <Field
              component={TextInput}
              endAdornment={<Search />}
              id="searcbar"
              name="search"
              placeholder="search"
            />

            <Button
              onClick={toggle}
              icon={open ? <TriangleDown /> : <TriangleUp />}
            >
              Menu
            </Button>
          </Form>
        </Formik>
      </div>
      <PhaseBanner className="masthead__phasebanner" />
      <div className="masthead__container">
        <Nav
          className={open ? 'is-open' : 'is-closed'}
          orientation="horizontal"
          navItems={navItems}
        />
      </div>
    </div>
  )
}

export default MastHead
