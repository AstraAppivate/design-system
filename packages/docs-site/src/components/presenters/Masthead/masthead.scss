@import "@royalnavy/css-framework";

$gutter: spacing(3);

.masthead__container {
  box-sizing: content-box;
  display: flex;
  flex-direction: column;
  max-width: $navy-content-width;
  margin: $gutter auto;
  padding: 0 $gutter;
}

.masthead__logo {
  width: 280px;
  margin-bottom: spacing(2);
}

.masthead__search {
  display: flex;
  flex-direction: row;
  margin: spacing(1) 0;
}

.masthead__search .rn-textinput {
  margin: 0;
}

.masthead__search .rn-btn {
  margin: 0 0 0 $gutter;
}

.masthead .rn-nav {
  display: none;
}

.masthead .rn-nav.is-open {
  display: block;
}

@include breakpoint("s") {
  .masthead__container {
    flex-direction: row;
    justify-content: space-between;
    flex-wrap: wrap;
  }

  .masthead__logo {
    width: 405px;
    margin-bottom: 0;
  }

  .masthead__search {
    margin: 0;
  }

  .masthead__search .rn-btn {
    display: none;
  }

  .masthead .rn-nav {
    display: block;
    margin-left: -1rem;
  }

  .masthead .rn-nav__list-item {
    display: inline-block;
    width: auto;
  }
}
