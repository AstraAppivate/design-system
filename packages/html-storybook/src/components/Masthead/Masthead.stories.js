import React from 'react'

import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import { withKnobs, text } from '@storybook/addon-knobs'

const stories = storiesOf('Masthead', module)

stories.addDecorator(withKnobs)

stories.add('Default', () => (
  <div class="rn-masthead h_mb-4">
    <div class="rn-masthead__head">
      <h4 class="rn-masthead__title">NELSON Apps</h4>
      <button
        class="rn-btn--primary neutral regular rn-masthead__action"
        type="button"
      >
        Action
      </button>
    </div>
    <nav class="rn-masthead__nav" />
  </div>
))

stories.add('With Links', () => (
  <div class="rn-masthead h_mb-4">
    <div class="rn-masthead__head">
      <h4 class="rn-masthead__title">NELSON Apps</h4>
      <button
        class="rn-btn--primary neutral regular rn-masthead__action"
        type="button"
      >
        Action
      </button>
    </div>
    <nav class="rn-masthead__nav">
      <a class="rn-masthead__link h_mr-1" href="/">
        Home
      </a>
      <a class="rn-masthead__link h_mr-1" href="/dashboard">
        Dashboard
      </a>
      <a class="rn-masthead__link h_mr-1" href="/settings">
        Settings
      </a>
    </nav>
  </div>
))
