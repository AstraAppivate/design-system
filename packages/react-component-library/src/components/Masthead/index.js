import PropTypes from 'prop-types'
import React, { Component } from 'react'
import Uuid from 'uuid'

class Masthead extends Component {
  static propTypes = {
    /** The content of the component */
    children: PropTypes.node,
    /** Specifies a custom classname for the compoennt */
    className: PropTypes.string,
    /** The title of the component */
    title: PropTypes.string,
    /** The links to use in the masthead */
    links: PropTypes.arrayOf(PropTypes.string),
  }

  static defaultProps = {
    children: '',
    className: '',
    title: '',
    links: [],
  }

  constructor(props) {
    super(props)
    this.state = {}
  }

  render() {
    const { title, className, links, children } = this.props
    return (
      <div className={`rn-masthead ${className}`}>
        <div className="rn-masthead__head">
          {title && <h4 className="rn-masthead__title">{title}</h4>}
          {children}
        </div>
        {links && (
          <nav className="rn-masthead__nav">
            {links.map(link => (
              <a
                className="rn-masthead__link h_mr-1"
                href={link.to}
                key={Uuid(link)}
              >
                {link.label}
              </a>
            ))}
          </nav>
        )}
      </div>
    )
  }
}

export default Masthead
