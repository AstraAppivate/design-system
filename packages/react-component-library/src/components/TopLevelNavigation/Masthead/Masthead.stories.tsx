import React from 'react'
import { Meta } from '@storybook/react/types-6-0'
import {
  IconChatBubble,
  IconExitToApp,
  IconPerson,
  IconSettings,
  IconHome,
} from '@royalnavy/icon-library'

import { Link } from '../../Link'
import {
  Masthead,
  MastheadNav,
  MastheadNavItem,
  MastheadUser,
  MastheadUserItem,
} from '.'
import { Notification, Notifications } from '../NotificationPanel'

export default {
  component: Masthead,
  subcomponents: {
    MastheadNav,
    MastheadNavItem,
    MastheadUser,
    MastheadUserItem,
    Notifications,
    Notification,
  },
  title: 'Masthead',
} as Meta

const notifications = (
  <Notifications link={<Link href="#" />}>
    <Notification
      link={<Link href="#" />}
      name="Thomas Stephens"
      action="added a new comment to your"
      on="review"
      when={new Date('2019-11-05T14:25:02.178Z')}
      description="At vero eos et accusamus et iusto odio dignissimos ducimus qui blanditiis praesentium voluptatum deleniti atque corrupti quos dolores"
    />
    <Notification
      link={<Link href="#" />}
      name="Thomas Stephens"
      action="added a new comment to your"
      on="review"
      when={new Date('2019-11-01T14:25:02.178Z')}
      description="At vero eos et accusamus et iusto odio dignissimos ducimus qui blanditiis praesentium voluptatum deleniti atque corrupti quos dolores"
    />
    <Notification
      link={<Link href="#" />}
      name="Thomas Stephens"
      action="added a new comment to your"
      on="review"
      when={new Date('2019-11-01T14:25:02.178Z')}
      description="At vero eos et accusamus et iusto odio dignissimos ducimus qui blanditiis praesentium voluptatum deleniti atque corrupti quos dolores"
    />
  </Notifications>
)

const user = (
  <MastheadUser initials="RN">
    <MastheadUserItem
      icon={<IconPerson />}
      link={<Link href="#">Profile</Link>}
    />
    <MastheadUserItem
      icon={<IconSettings />}
      link={<Link href="#">Settings</Link>}
    />
    <MastheadUserItem
      icon={<IconChatBubble />}
      link={<Link href="#">Support</Link>}
    />
    <MastheadUserItem
      icon={<IconExitToApp />}
      link={<Link href="#">Logout</Link>}
    />
  </MastheadUser>
)

export const Default = (props: any) => (
  <Masthead
    {...props}
    homeLink={<Link href="#" />}
    nav={
      <MastheadNav>
        <MastheadNavItem link={<Link href="#">Get started</Link>} isActive />
        <MastheadNavItem link={<Link href="#">Styles</Link>} />
        <MastheadNavItem link={<Link href="#">Components</Link>} />
        <MastheadNavItem link={<Link href="#">About</Link>} />
      </MastheadNav>
    }
    notifications={notifications}
    user={user}
  />
)

Default.args = {
  onSearch: (e: React.SyntheticEvent) => console.log,
  searchPlaceholder: 'Search',
  title: 'Royal Navy Design System',
  hasUnreadNotification: true,
}

export const CustomLogo = () => (
  <Masthead title="Royal Navy Design System" Logo={IconHome} />
)

CustomLogo.storyName = 'Custom logo'

export const WithoutLogo = () => (
  <Masthead title="Royal Navy Design System" hasDefaultLogo={false} />
)

WithoutLogo.storyName = 'Without logo'

export const WithSearch = () => (
  <Masthead
    onSearch={(term: string) => console.log}
    searchPlaceholder="Search..."
    title="Royal Navy Design System"
  />
)

WithSearch.storyName = 'With search'

export const WithAvatarLinks = () => (
  <Masthead title="Royal Navy Design System" user={user} />
)

WithAvatarLinks.storyName = 'With avatar links'

export const WithNotifications = () => (
  <Masthead title="Royal Navy Design System" notifications={notifications} />
)

WithNotifications.storyName = 'With notifications'

export const WithNavigation = () => (
  <Masthead
    homeLink={<Link href="#" />}
    title="Royal Navy Design System"
    nav={
      <MastheadNav>
        <MastheadNavItem link={<Link href="#">Get started</Link>} isActive />
        <MastheadNavItem link={<Link href="#">Styles</Link>} />
        <MastheadNavItem link={<Link href="#">Components</Link>} />
        <MastheadNavItem link={<Link href="#">About</Link>} />
      </MastheadNav>
    }
  />
)

WithNavigation.storyName = 'With navigation'
