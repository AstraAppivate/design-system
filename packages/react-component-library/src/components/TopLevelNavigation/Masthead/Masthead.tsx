import React, { useRef, useState } from 'react'

import { Bell, Logo as DefaultLogo, Search as SearchIcon } from '../../../icons'
import { LinkTypes } from '../../../common/Link'
import { MastheadUserProps } from './index'
import { Nav, NavItem } from '../../../common/Nav'
import {
  NOTIFICATION_CONTAINER_WIDTH,
  NotificationsProps,
} from '../NotificationPanel'
import { SHEET_PLACEMENT } from '../Sheet/constants'
import { Sheet } from '../Sheet/Sheet'
import { SheetButton } from '../Sheet/SheetButton'
import { useMastheadSearch } from './useMastheadSearch'
import { StyledOption } from './partials/StyledOption'
import { StyledOptions } from './partials/StyledOptions'
import { StyledMain } from './partials/StyledMain'
import { StyledMastHead } from './partials/StyledMasthead'
import { StyledServiceName } from './partials/StyledServiceName'
import { StyledTitle } from './partials/StyledTitle'
import { StyledVerticalSeparator } from './partials/StyledVerticalSeparator'
import { StyledBanner } from './partials/StyledBanner'
import { SearchBar } from '../SearchBar/SearchBar'
import { StyledNotRead } from '../NotificationPanel/partials/StyledNotRead'

export interface MastheadProps {
  hasDefaultLogo?: boolean
  hasUnreadNotification?: boolean
  homeLink?: React.ReactElement<LinkTypes>
  Logo?: React.ComponentType
  nav?: React.ReactElement<Nav<NavItem>>
  notifications?: React.ReactElement<NotificationsProps>
  onSearch?: (term: string) => void
  searchPlaceholder?: string
  title: string
  user?: React.ReactElement<MastheadUserProps>
}

function getServiceName(
  homeLink: React.ReactElement<LinkTypes>,
  Logo: React.ComponentType,
  title: string
) {
  const link = homeLink || <span />
  return React.cloneElement(link as React.ReactElement, {
    ...link.props,
    children: (
      <StyledServiceName>
        {Logo &&
          React.cloneElement(<Logo />, {
            role: 'presentation',
          })}
        <StyledTitle $hasLogo={!!Logo} data-testid="masthead-servicename">
          {title}
        </StyledTitle>
      </StyledServiceName>
    ),
  })
}

export const Masthead: React.FC<MastheadProps> = ({
  hasDefaultLogo = true,
  hasUnreadNotification,
  homeLink,
  Logo,
  nav,
  notifications,
  onSearch,
  searchPlaceholder = '',
  title,
  user,
  ...rest
}) => {
  const [showNotifications, setShowNotifications] = useState(false)
  const searchButtonRef = useRef<HTMLButtonElement>(null)

  const {
    containerWidth,
    mastheadContainerRef,
    setShowSearch,
    showSearch,
    toggleSearch,
  } = useMastheadSearch()

  const DisplayLogo = Logo ?? (hasDefaultLogo ? DefaultLogo : null)

  const submitSearch = (term: string) => {
    onSearch(term)
    setShowSearch(false)
  }

  return (
    <StyledMastHead
      $showNotifications={showNotifications}
      $showSearch={showSearch}
      data-testid="masthead"
      ref={mastheadContainerRef}
      {...rest}
    >
      <StyledMain>
        <StyledBanner data-testid="masthead-banner" role="banner">
          {getServiceName(homeLink, DisplayLogo, title)}
        </StyledBanner>

        <StyledOptions>
          {onSearch && (
            <>
              <StyledOption
                $isActive={showSearch}
                aria-expanded={showSearch}
                aria-label="Show search"
                as="button"
                onClick={toggleSearch}
                ref={searchButtonRef}
                data-testid="masthead-search-button"
              >
                <SearchIcon />
              </StyledOption>
              {user && (
                <StyledVerticalSeparator>
                  <line x1="0" y1="14" x2="0" y2="44" />
                </StyledVerticalSeparator>
              )}
            </>
          )}

          {notifications && (
            <Sheet
              button={(
                <StyledOption
                  aria-label="Show notifications"
                  as={SheetButton}
                  data-testid="notification-button"
                  icon={<Bell />}
                >
                  {hasUnreadNotification && (
                    <StyledNotRead data-testid="not-read" />
                  )}
                </StyledOption>
              )}
              placement={SHEET_PLACEMENT.BELOW}
              width={NOTIFICATION_CONTAINER_WIDTH}
              onHide={() => setShowNotifications(false)}
              onShow={() => setShowNotifications(true)}
            >
              {notifications}
            </Sheet>
          )}

          {(notifications || onSearch) && user && (
            <StyledVerticalSeparator>
              <line x1="0" y1="14" x2="0" y2="44" />
            </StyledVerticalSeparator>
          )}
          {user}
        </StyledOptions>
      </StyledMain>

      {onSearch && showSearch && (
        <SearchBar
          onSearch={submitSearch}
          searchButton={searchButtonRef}
          searchPlaceholder={searchPlaceholder}
          setShowSearch={setShowSearch}
          style={{ width: containerWidth }}
        />
      )}

      {nav}
    </StyledMastHead>
  )
}

Masthead.displayName = 'Masthead'
