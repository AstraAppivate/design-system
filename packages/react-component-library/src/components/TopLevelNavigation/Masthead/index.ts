export * from './Masthead'
export * from './MastheadNav'
export * from './MastheadNavItem'
export * from './MastheadUser'
export * from './MastheadUserItem'
