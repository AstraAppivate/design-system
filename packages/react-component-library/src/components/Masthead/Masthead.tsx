import React, { useRef, useState } from 'react'
import classNames from 'classnames'
import { CSSTransition, TransitionGroup } from 'react-transition-group'

import { Bell, Logo as DefaultLogo, Search as SearchIcon } from '../../icons'
import { MastheadUserProps } from '.'
import { Nav, NavItem } from '../../types/Nav'
import {
  NOTIFICATION_CONTAINER_WIDTH,
  NotificationsProps,
} from '../NotificationPanel'
import { Searchbar } from '../Searchbar'
import { SHEET_PLACEMENT } from '../TopLevelNavigation/constants'
import { Sheet } from '../TopLevelNavigation/Sheet'
import { SheetButton } from '../TopLevelNavigation/SheetButton'
import { useMastheadSearch } from './useMastheadSearch'

export interface MastheadProps {
  hasDefaultLogo?: boolean
  hasUnreadNotification?: boolean
  homeLink?: React.ReactElement<LinkTypes>
  Logo?: React.ComponentType
  nav?: React.ReactElement<Nav<NavItem>>
  notifications?: React.ReactElement<NotificationsProps>
  onSearch?: (term: string) => void
  searchPlaceholder?: string
  title: string
  user?: React.ReactElement<MastheadUserProps>
}

function getServiceName(
  homeLink: React.ReactElement<LinkTypes>,
  Logo: React.ComponentType,
  title: string
) {
  const link = homeLink || <span />

  return React.cloneElement(link as React.ReactElement, {
    ...link.props,
    children: (
      <>
        {Logo && <Logo />}
        <span className="rn-masthead__title" data-testid="masthead-servicename">
          {title}
        </span>
      </>
    ),
    className: classNames('rn-masthead__service-name', {
      'rn-masthead__service--has-logo': Logo,
    }),
  })
}

export const Masthead: React.FC<MastheadProps> = ({
  hasDefaultLogo = true,
  hasUnreadNotification,
  homeLink,
  Logo,
  nav,
  notifications,
  onSearch,
  searchPlaceholder = '',
  title,
  user,
}) => {
  const [showNotifications, setShowNotifications] = useState(false)
  const searchButtonRef = useRef<HTMLButtonElement>(null)

  const {
    containerWidth,
    mastheadContainerRef,
    setShowSearch,
    showSearch,
    toggleSearch,
  } = useMastheadSearch()

  const DisplayLogo = Logo ?? (hasDefaultLogo ? DefaultLogo : null)

  const classes = classNames('rn-masthead', {
    'rn-masthead--show-search': showSearch,
    'rn-masthead--show-notifications': showNotifications,
  })

  const searchButtonClasses = classNames(
    'rn-masthead__option',
    'rn-searchbar__btn',
    {
      'is-active': showSearch,
      'is-inactive': !showSearch,
    }
  )

  const submitSearch = (term: string) => {
    onSearch(term)
    setShowSearch(false)
  }

  return (
    <div className={classes} data-testid="masthead" ref={mastheadContainerRef}>
      <div className="rn-masthead__main">
        {getServiceName(homeLink, DisplayLogo, title)}

        <div className="rn-masthead__options">
          {onSearch && (
            <>
              <button
                className={searchButtonClasses}
                onClick={toggleSearch}
                ref={searchButtonRef}
                data-testid="masthead-search-button"
              >
                <SearchIcon />
              </button>
              <svg className="rn-masthead__vertical-separator">
                <line x1="0" y1="14" x2="0" y2="44" />
              </svg>
            </>
          )}

          {notifications && (
            <Sheet
              button={(
                <SheetButton
                  className="rn-masthead__option"
                  data-testid="notification-button"
                  icon={<Bell className="rn-sheet__icon" />}
                >
                  {hasUnreadNotification && (
                    <span
                      className="rn-notification-panel__not-read"
                      data-testid="not-read"
                    />
                  )}
                </SheetButton>
              )}
              className="rn-masthead__notification"
              placement={SHEET_PLACEMENT.BELOW}
              width={NOTIFICATION_CONTAINER_WIDTH}
              onHide={() => setShowNotifications(false)}
              onShow={() => setShowNotifications(true)}
            >
              {notifications}
            </Sheet>
          )}

          {user}
        </div>
      </div>

      <TransitionGroup>
        {onSearch && showSearch && (
          <CSSTransition
            classNames="rn-searchbar"
            timeout={{ enter: 300, exit: 300 }}
          >
            <Searchbar
              onSearch={submitSearch}
              searchButton={searchButtonRef}
              searchPlaceholder={searchPlaceholder}
              setShowSearch={setShowSearch}
              style={{ width: containerWidth }}
            />
          </CSSTransition>
        )}
      </TransitionGroup>

      {nav}
    </div>
  )
}

Masthead.displayName = 'Masthead'
