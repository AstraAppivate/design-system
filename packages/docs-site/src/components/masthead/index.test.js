import React from 'react'
import renderer from 'react-test-renderer'

import Masthead from './index'

describe('Masthead', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Masthead />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
