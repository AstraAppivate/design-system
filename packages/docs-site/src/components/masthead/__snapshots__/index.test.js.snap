// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Masthead renders correctly 1`] = `
<section
  className=" masthead"
>
  <div
    className="container"
  >
    <div
      className="masthead__content"
    />
  </div>
</section>
`;
